proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,374 @@
+
+- **It is not guaranteed to use the same layout as a C struct with a similar "shape".** If such a struct is necessary, it should be defined in a C header and imported into Swift.
+
+- **The fields are not guaranteed to be laid out in declaration order.** The compiler may choose to reorder fields to minimize padding while satisfying alignment requirements, for example, .
+
+That said, the compiler is allowed to use its knowledge of the struct's contents and layout to derive any of these properties. For instance, the compiler can statically prove that copying `Point` is "trivial", because each of its members has a statically-known type that is "trivial". However, depending on this at the language level is not supported, with two exceptions:
+
+1. The run-time memory layout of a struct with a single field is always identical to the layout of the instance property on its own, whether the struct is declared `@frozen` or not. This has been true since Swift 1. (This does not extend to the calling convention, however.)
+
+2. The representation of `nil` for any type that is "nullable" in C / Objective-C is the same as the representation of `nil` or `NULL` in those languages. This includes class references, class-bound protocol references, class type references, unsafe-pointers, `@convention(c)` functions, `@convention(block)` functions, OpaquePointer, Selector, and NSZone. This has been true since Swift 3 ([SE-0055][]).
+
+This proposal does not change either of these guarantees.
+
+  [trivial]: https://docs.microsoft.com/en-us/cpp/cpp/trivial-standard-layout-and-pod-types
+  [SE-0055]: https://github.com/apple/swift-evolution/blob/master/proposals/0055-optional-unsafe-pointers.md
+
+### `@frozen` on `enum` types
+
+Marking an enum as `@frozen` will also allow the compiler to optimize away runtime calls.
+
+In addition, marking an enum as frozen restores the ability of a library user to exhaustively switch over that enum without an `@unknown default:`, because it guarantees no further cases will be added.
+
+
+## Naming
+
+`@frozen` was used here, as originally used in [SE-0192][], for both fixed-layout structs and enums. 
+
+- While they share a fixing of their layout going forward, declaring an enum as frozen has additional meaning to _users_ of the library, whereas a frozen struct is an implementation detail of the library.
+
+- It would be reasonable to use `@frozen` to describe an enum without associated values that promises not to *add* any cases with associated values. This would allow for similar optimization as `@frozen` on a struct, while still not declaring that the enum is frozen.
+
+- Since this feature is only relevant for libraries with binary compatibility concerns, it may be more useful to tie its syntax to `@available` from the start, as with [SE-0193][]. In that case neither `@frozen` nor `@frozen` would be appropriate.
```

Comment:
We didn't actually do that in SE-0193, and find-and-replace mucked up this paragraph too.